core/src/main/java/com/linecorp/armeria/common/HttpRequest.java

```diff
@@ -280,6 +281,22 @@ static HttpRequest of(RequestHeaders headers, Publisher<? extends HttpObject> pu
         }
     }
 
+    /**
+     * Creates a new instance from an existing {@link RequestHeaders}, {@link Publisher} and trailers.
+     *
+     * <p>Note that the {@link HttpData}s in the {@link Publisher} are not released when
+     * {@link Subscription#cancel()} or {@link #abort()} is called. You should add a hook in order to
+     * release the elements. See {@link PublisherBasedStreamMessage} for more information.
+     */
+    static HttpRequest of(RequestHeaders headers,
+                          Publisher<? extends HttpData> publisher,
+                          HttpHeaders trailers) {
+        requireNonNull(headers, "headers");
+        requireNonNull(publisher, "publisher");
+        requireNonNull(trailers, "trailers");
+        return of(headers, new SurroundingPublisher<>(null, publisher, trailers));
```

Comment:
Can we add an optimization that wraps `publisher` only when `trailers` is not empty?

```
if (trailers.isEmpty()) {
  return of(headers, publisher);
} else { ... }
```

Reply:
✅ updated!